File: compiler/bench/toposort.c
// expected: 332833500
#include <stdio.h>
#define N 1000
int adj[N][5];
int deg[N];
int indegree[N];
int result[N];
int queue[N];

void edge(int from, int to) {
    adj[from][deg[from]++] = to;
    indegree[to]++;
}

void init_graph() {
    for (int i = 0; i < N; i++) {
        deg[i] = 0;
        indegree[i] = 0;
    }
    for (int i = 1; i < N; i++) {
        edge(i - 1, i);
    }
}

void toposort() {
    int qhead = 0, qtail = 0;
    for (int i = 0; i < N; i++) {
        if (indegree[i] == 0) {
            queue[qtail++] = i;
        }
    }
    int ridx = 0;
    while (qhead < qtail) {
        int node = queue[qhead++];
        result[ridx++] = node;
        for (int i = 0; i < deg[node]; i++) {
            int neighbor = adj[node][i];
            indegree[neighbor]--;
            if (indegree[neighbor] == 0) {
                queue[qtail++] = neighbor;
            }
        }
    }
}

int main() {
    init_graph();
    toposort();
    long sum = 0;
    for (int i = 0; i < N; i++) {
        sum += (long)result[i] * i;
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/toposort.fs
\ expected: 332833500
\ Topological sort on DAG of 1000 nodes, sum of sorted positions * node

1000 constant N
create adj N 5 * cells allot
create deg N cells allot
create indegree N cells allot
create result N cells allot
create queue N cells allot

: edge ( from to -- )
  over cells deg + @ 5 * rot + cells adj + !
  1 over cells indegree + +!
  swap cells deg + dup @ 1+ swap ! ;

: init-graph ( -- )
  N 0 do
    0 i cells deg + !
    0 i cells indegree + !
  loop
  N 1 do
    i 1- i edge
  loop ;

: toposort ( -- )
  0 0  \ qhead qtail
  N 0 do
    i cells indegree + @ 0= if
      i over cells queue + !
      swap 1+ swap
    then
  loop
  0  \ ridx
  begin 2 pick 2 pick < while
    3 pick cells queue + @
    4 roll 1+ 4 -roll
    dup 2 pick cells result + !
    rot 1+ -rot
    dup cells deg + @ 0 ?do
      over 5 * i + cells adj + @
      dup cells indegree + @ 1-
      dup 2 pick cells indegree + !
      0= if
        3 pick cells queue + !
        swap 1+ swap
      else drop then
    loop
    drop
  repeat
  drop 2drop ;

: checksum ( -- n )
  0 N 0 do
    i cells result + @ i * +
  loop ;

: main init-graph toposort checksum . cr ;
